src/services/WorkPeriodService.js

```diff
@@ -263,33 +259,33 @@ async function updateWorkPeriod (currentUser, id, data) {
   // check permission
   await _checkUserPermissionForWriteWorkPeriod(currentUser)
 
-  const workPeriod = await WorkPeriod.findById(id)
+  const workPeriod = await WorkPeriod.findById(id, { withPayments: true })
   const oldValue = workPeriod.toJSON()
-
-  // if resourceBookingId is provided then update projectId and userHandle
-  if (data.resourceBookingId) {
-    const resourceBooking = await helper.ensureResourceBookingById(data.resourceBookingId) // ensure resource booking exists
-    data.projectId = resourceBooking.projectId
-
-    const user = await helper.ensureUserById(resourceBooking.userId) // ensure user exists
-    data.userHandle = user.handle
+  if (oldValue.daysWorked === data.daysWorked) {
+    return oldValue
   }
-  // If one of the dates are missing then auto-calculate it
-  _autoCalculateDates(data)
-
-  data.updatedBy = await helper.getUserId(currentUser.userId)
-  let updated = null
-  try {
-    updated = await workPeriod.update(data)
-  } catch (err) {
-    if (!_.isUndefined(err.original)) {
-      throw new errors.BadRequestError(err.original.detail)
-    } else {
-      throw err
-    }
+  if (data.daysWorked < oldValue.daysPaid) {
+    throw new errors.BadRequestError(`Cannot update daysWorked (${data.daysWorked}) to the value less than daysPaid (${oldValue.daysPaid})`)
   }
-
-  await helper.postEvent(config.TAAS_WORK_PERIOD_UPDATE_TOPIC, updated.toJSON(), { oldValue: oldValue, key: `resourceBooking.id:${data.resourceBookingId}` })
+  const resourceBooking = await helper.ensureResourceBookingById(oldValue.resourceBookingId)
+  const weeks = helper.extractWorkPeriods(resourceBooking.startDate, resourceBooking.endDate)
+  if (_.isEmpty(weeks)) {
+    throw new errors.ConflictError('Resource booking has missing dates')
+  }
+  const thisWeek = _.find(weeks, ['startDate', oldValue.startDate])
+  if (_.isNil(thisWeek)) {
+    throw new errors.ConflictError('Work Period dates are not compatible with Resource Booking dates')
+  }
+  if (thisWeek.daysWorked < data.daysWorked) {
+    throw new errors.BadRequestError(`Maximum allowed daysWorked is (${thisWeek.daysWorked})`)
+  }
+  data.paymentStatus = helper.calculateWorkPeriodPaymentStatus(_.assign({}, oldValue, data))
+  if (oldValue.paymentStatus === data.paymentStatus) {
+    return oldValue
+  }
```

Comment:
@eisbilir if paymentStatus is not changed this logic would not update WP at all. But we can also update other values, like `daysWorked`, `daysPaid` and so on. Would not this cause issues?

Reply:
@eisbilir can we avoid all such places?

![image](https://user-images.githubusercontent.com/146016/122177295-25376f80-ce8e-11eb-9530-8281dbe27a35.png)

It feels like it would be better just update the records even though data is not changed. Would it break anything if I just remove such places?